src/audio/pipeline/pipeline-schedule.c

```diff
@@ -39,6 +39,139 @@ static void pipeline_schedule_cancel(struct pipeline *p)
+		if (err < 0) {
+			pipe_err(p, "pipeline_task(): xrun recover failed! pipeline will be stopped!");
+			/* failed - host will stop this pipeline */
+			return SOF_TASK_STATE_COMPLETED;
+		}
+	}
+
+	pipe_dbg(p, "pipeline_task() sched");
+
+	return SOF_TASK_STATE_RESCHEDULE;
+}
+
+/* DP pipeline pipeline task with Zephyr */
+static enum task_state pipeline_task_no_transition(struct pipeline *p)
+{
+	pipe_dbg(p, "pipeline_task(): no state %d", p->status);
+
+	/* work depends on the state */
+	switch (p->status) {
+	case COMP_STATE_PREPARE:
+	case COMP_STATE_SUSPEND:
+	case COMP_STATE_PAUSED:
+		/* sleep and wait for state changes */
+		k_sleep(K_MSECS(1));
+		return SOF_TASK_STATE_RESCHEDULE;
+	case COMP_STATE_ACTIVE:
+		return pipeline_active_task(p);
+	case COMP_STATE_INIT:
+	case COMP_STATE_READY:
+	default:
+		/* we should not get here with these states */
+		return SOF_TASK_STATE_COMPLETED;
+	}
+}
+
+/* current state is prepare, suspend or paused - handle change to next state */
+static enum task_state transition_from_inactive(struct pipeline *p)
+{
+	switch (p->status_next) {
+	case COMP_STATE_SUSPEND:
+	case COMP_STATE_PAUSED:
+	case COMP_STATE_PREPARE:
+		/* sleep and wait for state changes */
+		k_sleep(K_MSECS(1));
+		return SOF_TASK_STATE_RESCHEDULE;
+	case COMP_STATE_ACTIVE:
+		/* TODO: send trigger(start) */
+		p->status = p->status_next;
+		return SOF_TASK_STATE_RESCHEDULE;
+	case COMP_STATE_INIT:
+	case COMP_STATE_READY:
+	default:
+		/* we should not get here with these states */
+		return SOF_TASK_STATE_COMPLETED;
+	}
+}
+
+/* current state is active - handle change to next state */
+static enum task_state transition_from_active(struct pipeline *p)
+{
+	switch (p->status_next) {
+	case COMP_STATE_SUSPEND:
+	case COMP_STATE_PAUSED:
+	case COMP_STATE_PREPARE:
+		/* TODO: send trigger(stop) */
+		p->status = p->status_next;
+		return SOF_TASK_STATE_RESCHEDULE;
+	case COMP_STATE_ACTIVE:
+	case COMP_STATE_INIT:
+	case COMP_STATE_READY:
+	default:
+		/* we should not get here with these states */
+		return SOF_TASK_STATE_COMPLETED;
+	}
+}
+
+/* DP pipeline pipeline task with Zephyr */
+static enum task_state pipeline_task_transition(struct pipeline *p)
+{
+	pipe_dbg(p, "pipeline_task(): state %d", p->status);
+
+	/* work depends on the current state to next state */
+	switch (p->status) {
+	case COMP_STATE_PREPARE:
+	case COMP_STATE_SUSPEND:
+	case COMP_STATE_PAUSED:
+		return transition_from_inactive(p);
+	case COMP_STATE_ACTIVE:
+		return transition_from_active(p);
+	case COMP_STATE_INIT:
+	case COMP_STATE_READY:
+	default:
+		/* we should not get here with these states */
+		return SOF_TASK_STATE_COMPLETED;
+	}
+}
+/* DP pipeline pipeline task with Zephyr */
+static enum task_state pipeline_task(void *arg)
+{
+	struct pipeline *p = arg;
+
+	/* are we a state in transition */
+	if (p->status == p->status_next)
+		return pipeline_task_no_transition(p);
+	else
+		return pipeline_task_transition(p);
+}
+#else
```

Comment:
@lyakh @keyonjie see above pipeline state transition logic for dealing with trigger(start|stop|pause|release) and copy().
This could be refactored a little and support LL and xtos too. May help with a lot of our pause -> resume bugs